### DISCUSSION TITLE
A theoretical question about Lora strength

### DISCUSSION THREAD

**Discussion by FartyPants** [oobabooga/text-generation-webui#2033] [Ideas]
In SD we can tame overcooked LORAs by adding a strength after the training. It doesn't require any additional computation.

In theory - would the same idea be possible with LLM LORAs? To control their effect after training?

**Comment by cyberfox** [comment]
Greetings,
Yes, this is absolutely possible.  I fine-tuned a dataset using a rented 4090 and I overtrained it, unfortunately.  Results came back _verbatim_ from the dataset (there was some duplication in the data, and that ended up making those segments super-likely to come up) and it wasn't doing a great job of incorporating context, if it even tried.

In the `loras/{lora}/` directory, you'll find a pair of main files.  The parameters are in `adapter_model.bin`, but the interesting one for these purposes is `adapter_config.json`, which will contain a field:
```json
  "lora_alpha": 128,
```
and 128 was the alpha I set it to when training.  What I discovered was that before loading the LoRA, turning that down to (e.g.) 64 almost eliminated the LoRA's effect, and tuning it to around 96 gave a surprisingly good result.  It still was a little heavily tuned to the original dataset, and I haven't tried fine-tuning it much from there.  (Costs me ~$0.75/hour that I work on it because my local 3070Ti isn't up to handling the datasets.)

I'm going to spend more time tuning this, but **the tl;dr is: Yes, you can tweak the alpha at inference time, and it'll lower the strength of the LoRA**. This also means that you can mix and match LoRA's at different strengths, to create custom inference engines that are better at specific skills, while having others available but deprioritized.

I'm not sure if there's a limit to this; e.g. if by overtraining, I can make it _basically_ work, but it'll never be as good as a perfectly fine-tuned model, or if it's just as good as if I'd picked those values to start with.

I'm still learning; I hope this brain-dump helps! :)

--  Morgan

**Comment by mikolodz** [comment]
Hello @cyberfox !
Did you train on a instruction dataset, or using a txt plain file? I'm struggling a bit with a dataset containg hundreds of product specifications incl attributes, descriptions etc. There are some similar products in the dataset e.g. different sizes of the same product, although the quality of the dataset is rather high.
But i'm getting wrong responses from the model (wizard vicuna gptq 4b). For instance I ask about product X and getting description of another product. I even redesigned the dataset so that it would contain product name in every piece of information, so for example in the specification section it says "Product X specification:". This way I made sure that the model will be aware of to which product does the specification belong to. Didn't help at all.

I'm bit disappointed. I have tried training using both Lora alpha value = 64 and 256. Similar outcome - rubbish. Previously I've done some training on other datasets like sensmaking.json (which worked properly), alpaca_pl.json (to train the polish language - it resulted in some dual-language responses which was somewhat disturbing), but this time I have a high quality database/documentation and I'm getting really poor results.

Maybe you have some observation or idea how could I train on such set?
Thanks!

**Comment by cyberfox** [comment]
@mikolodz I'm so sorry, I don't have any notifications set up so I missed this.

I have been doing training using raw text files, not anything structured.  I'm trying to
1. improve document awareness for a corpus of new documents (source code and design/architecture documentation), and
2. improve chat behavior for roleplaying purposes.

I've had some good luck with training chat behavior, but the document corpus has some weird behavior. Sometimes it'll completely copy data out of the documents, and relevant information in fact, but then it dies and goes into a weird PHP dump when it runs out of local context. (I think I've figured out why that happens, and it's an oobabooga bug, but a hard one to fix.)

I've had...problems with doing training sometimes; most recently I had a run where _every single data file_ in the `adapter_model.bin` zip file was just filled with nul bytes (0s). It pissed me off so much, because I ran it for 2 hours, and got no errors, no console messages...nothing to indicate that it was a complete failure.

That said, my data is unstructured, and I'm just trying to improve its conception of specialized knowledge, rather than very specific product lookups.

For your kind of data, very structured, multi-attributed data, you may want to implement fine-tuning yourself, instead of using the webui's version.  To be honest, I'm not sure you wouldn't be better served by using a vector embedding database (essentially the same kind of thing as SuperBooga is using) to look up vector embeddings that are similar to your query, and then inject that description into the prompt, along with your prompt.

You might want to look into the [LangChain](https://www.youtube.com/watch?v=2xxziIWmaSA&t=1734s) stuff. I personally don't use it, because I'm trying to do something a bit different, but it has tools for chunking your data, storing it in vector embedding DBs, doing similarity lookup from your query, pulling the chunks in, and passing them to an LLM.

So for your situation I'd say the first, and most important thing, is to ask...**_what kind of loss were you seeing_**?  It should have been logging the loss every _some number_ of batches, and you should see the loss trending steadily downward.  If you see loss spiking up, that's a problem, for instance.  Even if it continues trending down, but the loss is high in absolute terms, that'll give you an idea that it's not doing a great job predicting the next token.

tl;dr - Without knowing the shape of your data, I'm not sure what else to recommend, though.  Start with the loss, and if that seems okay, but the data is still wrong, you might try a vector embedding db, and see if that retrieves information in a way that is useful to you.

**Comment by CodeofDutyAI** [comment]
@cyberfox hi, can you please provide a reference on how this alpha value can be adjusted during inference? I couldn't understand that from a mathematical perspective.
